subprojects/composite-builds/src/main/java/org/gradle/composite/internal/CompositeBuildDependencySubstitutions.java

```diff
@@ -74,12 +74,16 @@ private ProjectComponentIdentifier getReplacementFor(ModuleComponentSelector sel
         ModuleIdentifier candidateId = selector.getModuleIdentifier();
         Collection<ProjectComponentIdentifier> providingProjects = replacementMap.get(candidateId);
         if (providingProjects.isEmpty()) {
-            LOGGER.debug("Found no composite build substitute for module '" + candidateId + "'.");
+            if (LOGGER.isDebugEnabled()) {
+                LOGGER.debug("Found no composite build substitute for module '" + candidateId + "'.");
```

Comment:
Is it not simpler to use: `LOGGER.debug("Found no composite build substitute for module '{}'.", candidateId);

Reply:
It is but that's not equivalent though, because _if_ there's logging, the formatter will have to parse the pattern and substitute, which is using something similar to what `String.format` does.

Reply:
Sorry, this is a pet peeve of mine. We should do what @ljacomet suggests, because:

- parsing the {} is not as slow as `String.format()`, but much faster,
- by adding the `if` now you check the log level twice when debug logging is enabled, which needs to go through synchronization etc, and can actually make this slower than the `if`-less variant,
- performance is irrelevant, since we will only parse the string when debug logging is enabled.

The bottom line is: let's use the simpler approach. The main point is to avoid the string concatenation, and that will work with using `{}`.

Reply:
I changed it. As for whether performance is relevant or not, for the debug case, logging is probably slowing the build already, so no big deal. For the info use case below, that's more a concern, but the situation won't be as bad as it is today.